### ISSUE TITLE
Inconsistent returns from "this" and its fields

### ISSUE THREAD

**Issue by twolodzko** [brimdata/super#4585]
Consider the examples:

```shell
$ echo {1..3} | zq 'yield [this, this+1] | this' -
$ echo {1..3} | zq 'yield [this, this+1] | this[0]' -
1
2
3
$ echo {1..3} | zq 'yield {fst:this, snd:this+1} | this' -   
$ echo {1..3} | zq 'yield {fst:this, snd:this+1} | this["fst"]' -
1
2
3
$ echo {1..10} | zq 'yield {fst:this, snd:this+1} | this.fst' -   
$ echo {1..3} | zq 'yield {fst:this, snd:this+1} | yield this["fst"]' -
1
2
3
```
In some cases using `yield` is needed, in some not.

**Comment by mccanne** [comment]
Thanks @twolodzko!  We need to do some work here to create clearer rules for implied search vz implied yield.  You can always resort to explicit operator commands but there are lots of situations when implied operators can be a bit confusing so we should rationalize it better.  Hope to have a look at this soon...

**Comment by philrz** [comment]
@twolodzko: Since you seem like an advanced Zed user, you may already have found these docs/commands, but here's a couple things I wanted to make sure you're aware of.

1. @mccanne mentioned "implied `yield`", and to that end, the docs at https://zed.brimdata.io/docs/language/overview#26-implied-operators talk about the rules for how shorthand syntax is evaluted.

2. `zq -C` is a handy command that will show your Zed in "canonical" form, i.e., the parser will resolve any shorthand and therefore reflect how the query was ultimately executed. So taking a couple of your examples, this illustrates how the bare `this` was treated as a search.

```
$ zq -C 'yield [this, this+1] | this'
yield [this,this+1]
| search this

$ zq -C 'yield [this, this+1] | this[0]'
yield [this,this+1]
| yield this[0]
```

As @mccanne noted, your experience illustrates that we're probably due to have a rethink in this area, so we'll circle back here with more comments when we've had some internal discussions.

**Comment by twolodzko** [comment]
Thanks @philrz, will look at it.

**Comment by philrz** [comment]
We've discussed this internally and the current thinking is that we should enhance the PEG grammar with a list of reserved words (such as `this`) that can't be used "bare" in searches, since that way in the `| this` examples above there'd be a helpful error message rather than null output. And when a user wants to search for such a word, it would have to be quoted or escaped, though the choice between those two is TBD.

**Comment by philrz** [comment]
Verified in super commit 5a0232a.

This was addressed in #4585 by now requiring the explicit `?` (or the `search` keyword) to signal the keyword search operator, which means that there's no longer the unintended treatment of  the bare`| this` as a search. Revisiting the examples that kicked off the issue, we now see a helpful error message rather than a lack of output in those cases.

```
$ super -version
Version: v1.18.0-119-g5a0232a1

$ echo {1..3} | super -c 'yield [this, this+1] | this' -
parse error at line 1, column 28:
yield [this, this+1] | this
                       === ^ ===

$ echo {1..3} | super -c 'yield [this, this+1] | this[0]' -
1
2
3

$ echo {1..3} | super -c 'yield {fst:this, snd:this+1} | this' -
parse error at line 1, column 36:
yield {fst:this, snd:this+1} | this
                               === ^ ===

$ echo {1..3} | super -c 'yield {fst:this, snd:this+1} | this["fst"]' -
1
2
3

$ echo {1..3} | super -c 'yield {fst:this, snd:this+1} | this.fst' -
1
2
3

$ echo {1..3} | super -c 'yield {fst:this, snd:this+1} | yield this["fst"]' -
1
2
3
```

If the user had _wanted_ to search for the word `this`, they could use the `?`, but in that original input data it would show no output as before.

```
$ echo {1..3} | super -c 'yield [this, this+1] | ? this' -
[no output]

$ echo '"What is this?" "What is that?"' | super -c 'yield [this, this+1] | ? this' -
["What is this?",error("incompatible types")]
```

Thanks @mccanne!
